fairscale/nn/misc/grad_bucket.py

```diff
@@ -0,0 +1,133 @@
+from typing import Any, Callable, List, Optional, Union
+
+import torch
+
+
+class GradBucket:
+    """
+    Helper class to simplify the handling of gradient buckets
+    """
+
+    def __init__(self, size: int, dtype: torch.dtype, device: torch.device, destination: int) -> None:
+        self._max_size = size
+        self._params: List[torch.Tensor] = []
+        self._fill = 0
+        self._is_collapsed = False
+
+        # The actual flat tensor
+        self.buffer: Optional[torch.Tensor] = torch.zeros(self._max_size, dtype=dtype, device=device)
+
+        self.params_checked_in = 0
+        self.destination = destination
+        self.sent = True
+        self.callback: Optional[Callable[[Any], None]] = None
+
+    def reset_checked_in(self) -> None:
+        """ Reset the counter of the parameter grads which have been checked in
+        """
+        self.params_checked_in = 0
+        self.sent = False
+
+    @property
+    def all_checked_in(self) -> bool:
+        """ Have all the expected gradient check-in happened ?"""
+        return len(self._params) == self.params_checked_in
+
+    def can_add_grad_view(self, param: torch.Tensor) -> bool:
+        """ Is there enough room in the bucket to add this parameter gradient ?
+        """
+        return self._fill + param.numel() < self._max_size
+
+    def to(  # type: ignore
+        self,
+        device: Optional[Union[int, torch.device]],
+        dtype: Optional[torch.dtype] = None,
+        non_blocking: bool = False,
+    ) -> "GradBucket":
+        """
+        Move the underlying buffer
+        """
+        if self._is_collapsed:
+            self.rebuild()
+
+        assert self.buffer is not None, "Cannot move a collapsed bucket, please rebuild it"
+        self.buffer.to(device, dtype, non_blocking)
+
+    def zero(self) -> None:
+        """
+        Set all the grads to zero
+        """
+        if self.buffer is not None:
+            self.buffer.fill_(0.0)
+
+    @torch.no_grad()
+    def add_grad(self, param: torch.Tensor) -> None:
+        """
+        Add a new parameter gradient to the bucket. Param.grad becomes a view of this bucket buffer
+        """
+        if param.grad is None:
+            param.grad = torch.zeros_like(param)
```

Comment:
@zhengwy888 the grad needs to be a "real" tensor, else .zero_grad() calls fail because grad.detach_() will fail (see https://pytorch.org/docs/stable/autograd.html?highlight=detach_#torch.Tensor.detach_)